mqtt-cs/src/main/java/org/apache/rocketmq/mqtt/cs/session/infly/PushAction.java

```diff
@@ -181,6 +185,12 @@ public void write(Session session, Message message, int mqttId, int qos, Subscri
                 data = MqttMessageFactory.buildPublishMessage(topicName, message.getPayload(), qos, retained, mqttId);
                 break;
             case MQTT_5:
+                // Server send PublishMessage flow control
+                if (qos > 0 && !channelManager.publishSendTryAcquire(channel)) {
+                    logger.error("publishSendTryAcquire failed, client:{}, message:{}", clientId, message);
+                    return;
+                }
```

Comment:
If the push fails, there will be a retry, and the retry also requires flow control.

Reply:
Yes, the retry logic will also step into the `write` function and require the quota by flow control.